components/post_view/combined_system_message/combined_system_message.jsx

```diff
@@ -0,0 +1,277 @@
+            if (
+                postType === Posts.POST_TYPES.ADD_TO_TEAM ||
+                postType === Posts.POST_TYPES.ADD_TO_CHANNEL
+            ) {
+                Object.entries(values).forEach(([actorId, otherUserIds]) => {
+                    userIds = [...userIds, ...otherUserIds, actorId];
+                });
+            } else {
+                userIds = [...userIds, ...values];
+            }
+        });
+
+        const missingProfiles = await this.props.actions.getMissingProfilesByIds(userIds);
+
+        // trick to load missing profiles
+        this.setState({missingProfiles});
+    }
+
+    getDisplayNameById = (userIds = []) => {
+        const {
+            currentUserId,
+        } = this.props;
+
+        const displayNames = userIds.
+            filter((userId) => {
+                return userId !== currentUserId;
+            }).
+            map((userId) => {
+                return getDisplayNameByUserId(userId);
+            });
+
+        const userInProp = userIds.includes(currentUserId);
+        if (userInProp) {
+            displayNames.unshift(localizeMessage('combined_system_message.you', 'You'));
+        }
+
+        return displayNames;
+    }
+
+    generateFormattedMessage(postType, userDisplayNames, actor) {
+        const firstUser = userDisplayNames[0];
+        const numOthers = userDisplayNames.length - 1;
+
+        let formattedMessage;
+        if (numOthers === 0) {
+            formattedMessage = (
+                <FormattedHTMLMessage
+                    id={combinedSystemMessage[postType].one.id}
+                    defaultMessage={combinedSystemMessage[postType].one.defaultMessage}
+                    values={{
+                        firstUser,
+                        actor,
+                    }}
+                />
+            );
+        } else if (numOthers === 1) {
+            formattedMessage = (
+                <FormattedHTMLMessage
+                    id={combinedSystemMessage[postType].two.id}
+                    defaultMessage={combinedSystemMessage[postType].two.defaultMessage}
+                    values={{
+                        firstUser,
+                        secondUser: userDisplayNames[1],
+                        actor,
+                    }}
+                />
+            );
+        } else {
+            formattedMessage = (
+                <LastUsers
+                    actor={actor}
+                    expandedLocale={combinedSystemMessage[postType].many_expanded}
+                    postType={postType}
+                    userDisplayNames={userDisplayNames}
+                />
+            );
+        }
+
+        return formattedMessage;
+    }
+
+    render() {
+        const {
+            currentUserId,
+            userActivityProps,
+        } = this.props;
+
+        const messages = [];
+        Object.entries(userActivityProps).forEach(([postType, values]) => {
```

Comment:
This is almost identical to the `getMissingUserIds` stuff, I wonder if there is a better way to do this to only run this once instead, probably better for performance cause you have 3 iterations in here